src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+    /// value is updated, and the old value is returned.
+    pub(crate) fn insert(&mut self, key: String, count: usize) -> Option<usize> {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(map) => map.insert(key, count),
+            MultiThreaded(arc) => arc.insert(key, count),
+        }
+    }
+
+    /// Inserts a key-count pair into the map if the key does not already exist.
+    /// If the key does already exist, adds count to the current value of the
+    /// existing key.
+    pub(crate) fn insert_or_add(&mut self, key: String, count: usize) {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(map) => *map.entry(key).or_insert(0) += count,
+            MultiThreaded(arc) => arc.upsert(key, || count, |v| *v += count),
+        }
+    }
+
+    pub(crate) fn is_concurrent(&self) -> bool {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(_) => false,
+            MultiThreaded(_) => true,
+        }
+    }
+
+    pub(crate) fn write_and_clear<W>(&mut self, mut writer: W) -> io::Result<()>
+    where
+        W: io::Write,
+    {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(ref mut map) => {
+                let mut contents: Vec<_> = map.drain().collect();
+                contents.sort();
+                for (key, value) in contents {
+                    writeln!(writer, "{} {}", key, value)?;
+                }
+            }
+            MultiThreaded(ref mut arc) => {
+                let map = match Arc::get_mut(arc) {
+                    Some(map) => map,
+                    None => panic!(
+                        "Attempting to drain the contents of a concurrent HashMap \
+                         when more than one thread has access to it, which is \
+                         not allowed."
+                    ),
+                };
+                let map = mem::replace(map, CHashMap::with_capacity(CAPACITY_HASHMAP));
+                let mut contents: Vec<_> = map.into_iter().collect();
+                contents.sort();
+                for (key, value) in contents {
+                    writeln!(writer, "{} {}", key, value)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Demangles partially demangled Rust symbols that were demangled incorrectly by profilers like
+/// `sample` and `DTrace`.
+///
+/// For example:
+///     `_$LT$grep_searcher..searcher..glue..ReadByLine$LT$$u27$s$C$$u20$M$C$$u20$R$C$$u20$S$GT$$GT$::run::h30ecedc997ad7e32`
+/// becomes
+///     `<grep_searcher::searcher::glue::ReadByLine<'s, M, R, S>>::run`
+///
+/// Non-Rust symobols, or Rust symbols that are already demangled, will be returned unchanged.
+///
+/// Based on code in https://github.com/alexcrichton/rustc-demangle/blob/master/src/legacy.rs
+#[allow(clippy::cognitive_complexity)]
+pub(crate) fn fix_partially_demangled_rust_symbol(symbol: &str) -> Cow<str> {
```

Comment:
I assume this hasn't been changed from what's in `master`, just moved?

Reply:
Yes - That's correct.